crates/fm/src/lib.rs

```diff
@@ -10,28 +10,42 @@ pub use file_map::{File, FileId, FileMap};
 
 use std::{
     collections::HashMap,
-    path::{Path, PathBuf},
+    path::{Component, Path, PathBuf},
 };
 
 pub const FILE_EXTENSION: &str = "nr";
 
 #[derive(Clone, Debug, PartialEq, Eq, Hash)]
 struct VirtualPath(PathBuf);
 
-#[derive(Debug, Default)]
+#[derive(Debug)]
 pub struct FileManager {
+    root: PathBuf,
     file_map: file_map::FileMap,
     id_to_path: HashMap<FileId, VirtualPath>,
     path_to_id: HashMap<VirtualPath, FileId>,
 }
 
 impl FileManager {
-    // XXX: Maybe use a AsRef<Path> here, for API ergonomics
-    pub fn add_file(&mut self, path_to_file: &Path) -> Option<FileId> {
+    pub fn new(root: &Path) -> Self {
+        Self {
+            root: normalize_path(root),
+            file_map: Default::default(),
+            id_to_path: Default::default(),
+            path_to_id: Default::default(),
+        }
+    }
+
+    pub fn add_file(&mut self, file_name: &Path) -> Option<FileId> {
         // Handle both relative file paths and std/lib virtual paths.
-        let base = Path::new(".").canonicalize().expect("Base path canonicalize failed");
-        let res = path_to_file.canonicalize().unwrap_or_else(|_| path_to_file.to_path_buf());
-        let resolved_path = res.strip_prefix(base).unwrap_or(&res);
+        let absolute_path = normalize_path(&self.root.join(file_name));
+        let resolved_path = if file_name.starts_with("std/") || file_name.starts_with("std\\") {
```

Comment:
It would be nice to create a `is_stdlib_path` function for this but can be handled as part of the TODO.